storm-server/src/main/java/org/apache/storm/scheduler/resource/strategies/scheduling/sorter/NodeSorterHostProximity.java

```diff
@@ -0,0 +1,768 @@
+     * calculation, nodes nodes that have more balanced resource availability. So we will be less likely to pick a node that have a lot of
+     * one resource but a low amount of another.
+     *
+     * @param availHosts a collection of all the hosts we want to sort
+     * @param rackId     the rack id availNodes are a part of
+     * @return an iterable of sorted hosts.
+     */
+    private Iterable<ObjectResourcesItem> sortHosts(
+            Collection<String> availHosts, ExecutorDetails exec, String rackId,
+            Map<String, AtomicInteger> scheduledCount) {
+        ObjectResourcesSummary rackResourcesSummary = new ObjectResourcesSummary("RACK");
+        availHosts.forEach(h -> {
+            ObjectResourcesItem hostItem = new ObjectResourcesItem(h);
+            for (RasNode x : hostnameToNodes.get(h)) {
+                hostItem.add(new ObjectResourcesItem(x.getId(), x.getTotalAvailableResources(), x.getTotalResources(), 0, 0));
+            }
+            rackResourcesSummary.addObjectResourcesItem(hostItem);
+        });
+
+        LOG.debug(
+                "Rack {}: Overall Avail [ {} ] Total [ {} ]",
+                rackId,
+                rackResourcesSummary.getAvailableResourcesOverall(),
+                rackResourcesSummary.getTotalResourcesOverall());
+
+        return sortObjectResources(
+            rackResourcesSummary,
+            exec,
+            (hostId) -> {
+                AtomicInteger count = scheduledCount.get(hostId);
+                if (count == null) {
+                    return 0;
+                }
+                return count.get();
+            });
+    }
+
+    /**
+     * Nodes are sorted by two criteria.
+     *
+     * <p>1) the number executors of the topology that needs to be scheduled is already on the node in
+     * descending order. The reasoning to sort based on criterion 1 is so we schedule the rest of a topology on the same node as the
+     * existing executors of the topology.
+     *
+     * <p>2) the subordinate/subservient resource availability percentage of a node in descending
+     * order We calculate the resource availability percentage by dividing the resource availability that have exhausted or little of one of
+     * the resources mentioned above will be ranked after on the node by the resource availability of the entire rack By doing this
+     * calculation, nodes nodes that have more balanced resource availability. So we will be less likely to pick a node that have a lot of
+     * one resource but a low amount of another.
+     *
+     * @param availRasNodes a list of all the nodes we want to sort
+     * @param hostId     the host-id that availNodes are a part of
+     * @return an {@link Iterable} of sorted {@link ObjectResourcesItem} for nodes.
+     */
+    private Iterable<ObjectResourcesItem> sortNodes(
+            List<RasNode> availRasNodes, ExecutorDetails exec, String hostId,
+            Map<String, AtomicInteger> scheduledCount) {
```

Comment:
Can you please help me understand why AtomicInteger is required? My understanding is that scheduling is done sequentially in one thread.

Reply:
Easier to handle incrementing with a mutable Integer: otherwise a single statement becomes a three liner  with immutable Integer - get, increment, set.

Reply:
It feels like an overuse of AtomicInteger.  If this is only about reducing the code, I would suggest to use getOrDefault method, for example,

```
perNodeScheduledCount.computeIfAbsent(superId, (sid) -> new AtomicInteger(0))
                        .getAndAdd(entry.getValue().size());
```
can be changed to 

```
perNodeScheduledCount.put(superId, perNodeScheduledCount.getOrDefault(superId, 0) + entry.getValue().size());
```
And 
```
(rackId) -> {
                AtomicInteger count = scheduledCount.get(rackId);
                if (count == null) {
                    return 0;
                }
                return count.get();
            }
```
can be changed to 
```
(rackId) -> scheduledCount.getOrDefault(rackId, 0)
```

Reply:
I have seen that use case. First idiom is most common (but ugly) with the put and get smashed into one. Gets you part of the way but not fully like a mutable-int will do. Less common is use of int[1] but some go as far as to create a class. I believe there is a MutableInt class in storm project that is not used anywhere too.

AtomicInteger is the easier and best alternative (in my opinion) for using any mutable integer - because:
   (1) It is mutable
   (2) it is available in Java

I don't understand what you mean by overuse.  I think String and int are probably most overused in Java.

Reply:
But if you strongly object to use of AtomicInteger let me know. I will revert back to regular int/Integer

Reply:
I would still suggest against the use of AtomicInteger here. 

If you think the first implementation is ugly, you can separate them to two lines.
```
int curCnt = perNodeScheduledCount.getOrDefault(superId, 0);
perNodeScheduledCount.put(superId, curCnt + entry.getValue().size());
```
`perNodeScheduledCount` is a map and increasing the value of the map is a normal practice. 

And the second implementation also looks simpler comparing to AtomicInteger.

What I meant by overuse of AtomicInteger is when people (at least me) look at AtomicInteger, they wonder if there is concurrency going on and they have to think twice before making any changes to it.

AtomicInteger has int value inside the object (if you since int is overused)

Reply:
Although I still have different opinions after offline discussion, I won't block this PR on this.

Reply:
I may try int[] and see if that looks any better.